edge-modules/azure-monitor/src/Program.cs

```diff
@@ -51,8 +51,11 @@ static async Task<int> MainAsync()
             ModuleClient moduleClient = null;
             try
             {
-                moduleClient = await ModuleClient.CreateFromEnvironmentAsync(transportSettings);
-                moduleClient.ProductInfo = Constants.ProductInfo;
+                ModuleClientWrapper moduleClientWrapper = await ModuleClientWrapper.BuildModuleClientWrapperAsync(transportSettings);
+                SemaphoreSlim moduleClientLock = new SemaphoreSlim(1, 1);
+
+                IothubConnectionManager iothubConnectionManager = new IothubConnectionManager(moduleClientWrapper, moduleClientLock);
```

Comment:
The semaphore is passed into the same objects as ModuleClientWrapper, semaphore.WaitAsync is always called immediately before calling into ModuleClientWrapper, and semaphore.Release is always called immediately after calling into ModuleClientWrapper. The semaphore should probably be a private property of the wrapper, so the caller doesn't have to manage it. That way, you also don't have to expose ModuleClientWrapper.Inner.

Reply:
I agree with this, however I thought it was a more flexible abstraction if we allowed user access to the underlying `ModuleClient` so that they can do whatever they want with it.

I don't like how the user has to manage a separate lock though. In retrospect I should have made the lock a publicly exposed member of the abstraction. So we can either:
1. Make the lock a publicly exposed member of the abstraction
2. Write methods in the abstraction for every case where we use the `ModuleClient`

I prefer option 1, however if you prefer 2 I will go that route.

Reply:
Discussed with Damon and he convinced me option 2 is the right way. Made the changes.